public/app/features/plugins/components/AppRootPage.tsx

```diff
@@ -19,118 +37,160 @@ interface Props extends GrafanaRouteComponentProps<RouteParams> {}
 
-  shouldComponentUpdate(nextProps: Props) {
-    return nextProps.location.pathname.startsWith('/a/');
-  }
+  useEffect(() => {
+    loadAppPlugin(match.params.pluginId, dispatch);
+  }, [match.params.pluginId]);
 
-  async loadPluginSettings() {
-    const { params } = this.props.match;
-    try {
-      const app = await getPluginSettings(params.pluginId).then((info) => {
-        const error = getAppPluginPageError(info);
-        if (error) {
-          appEvents.emit(AppEvents.alertError, [error]);
-          this.setState({ nav: getWarningNav(error) });
-          return null;
-        }
-        return importAppPlugin(info);
-      });
-      this.setState({ plugin: app, loading: false, nav: undefined });
-    } catch (err) {
-      this.setState({
-        plugin: null,
-        loading: false,
-        nav: process.env.NODE_ENV === 'development' ? getExceptionNav(err) : getNotFoundNav(),
-      });
-    }
-  }
+  const onNavChanged = useCallback(
+    (newPluginNav: NavModel) => dispatch(stateSlice.actions.changeNav(newPluginNav)),
+    []
+  );
 
-  componentDidMount() {
-    this.loadPluginSettings();
+  if (!plugin || match.params.pluginId !== plugin.meta.id) {
+    return <Page navId="apps">{loading && <PageLoader />}</Page>;
   }
 
-  componentDidUpdate(prevProps: Props) {
-    const { params } = this.props.match;
-
-    if (prevProps.match.params.pluginId !== params.pluginId) {
-      this.setState({ loading: true, plugin: null });
-      this.loadPluginSettings();
-    }
+  if (!plugin.root) {
+    return (
+      <Page navId="apps">
+        <div>No root app page component found</div>;
+      </Page>
+    );
   }
 
-  onNavChanged = (nav: NavModel) => {
-    this.setState({ nav });
-  };
-
-  render() {
-    const { loading, plugin, nav, portalNode } = this.state;
+  const pluginRoot = plugin.root && (
+    <plugin.root
+      meta={plugin.meta}
+      basename={match.url}
+      onNavChanged={onNavChanged}
+      query={queryParams as KeyValue}
+      path={location.pathname}
+    />
+  );
+
+  if (config.featureToggles.topnav && !pluginNav) {
+    return <PluginPageContext.Provider value={context}>{pluginRoot}</PluginPageContext.Provider>;
+  }
 
-    if (!plugin || this.props.match.params.pluginId !== plugin.meta.id) {
-      return (
+  return (
+    <>
+      <InPortal node={portalNode}>{pluginRoot}</InPortal>
+      {sectionNav ? (
+        <Page navModel={sectionNav} pageNav={pluginNav?.node}>
+          <Page.Contents isLoading={loading}>
+            <OutPortal node={portalNode} />
+          </Page.Contents>
+        </Page>
+      ) : (
         <Page>
-          <PageLoader />
+          <OutPortal node={portalNode} />
         </Page>
-      );
-    }
+      )}
+    </>
+  );
+}
 
-    if (!plugin.root) {
-      // TODO? redirect to plugin page?
-      return <div>No Root App</div>;
+function buildPluginSectionNav(location: H.Location, pluginNav: NavModel | null, navIndex: NavIndex) {
```

Comment:
could probably pull this + couple of the functions below out into a utils file? we can do that later though i think

Reply:
yea, can move this out and add unit test for it